src/lib/paginator/paginator.ts

```diff
@@ -0,0 +1,166 @@
+import {MATERIAL_COMPATIBILITY_MODE} from '../core';
+
+/**
+ * Change event object that is emitted when the user selects a
+ * different page length or navigates to another page.
+ */
+export class PageChangeEvent {
+  currentPageIndex: number;
+  pageLength: number;
+  listLength: number;
+}
+
+/**
+ * Component to provide navigation between paged information. Displays the length of the current
+ * page, user-selectable options to change that length, what items are being shown, and
+ * navigational button to go to the previous or next page.
+ */
+@Component({
+  moduleId: module.id,
+  selector: 'md-paginator, mat-paginator',
+  templateUrl: 'paginator.html',
+  styleUrls: ['paginator.css'],
+  host: {
+    'class': 'mat-paginator',
+  },
+  providers: [
+    {provide: MATERIAL_COMPATIBILITY_MODE, useValue: false}
+  ],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  encapsulation: ViewEncapsulation.None,
+})
+export class MdPaginator implements OnInit {
+  private _initialized: boolean;
+
+  /** The zero-based current page index of the displayed list of items. Defaulted to 0. */
+  @Input() currentPageIndex: number = 0;
+
+  /** The list length of the total number of items that are being paginated. Defaulted to 0. */
+  @Input() listLength: number = 0;
+
+  /** Number of items to display on a page. By default set to the first of page length option. */
+  @Input()
+  set pageLength(pageLength: number) {
+    this._pageLength = pageLength;
+    this._updateDisplayedPageLengthOptions();
+  }
+  get pageLength(): number { return this._pageLength; }
+  _pageLength: number;
+
+  /** The set of provided page length options to display to the user. */
+  @Input()
+  set pageLengthOptions(pageLengthOptions: number[]) {
+    this._pageLengthOptions = pageLengthOptions;
+    this._updateDisplayedPageLengthOptions();
+  }
+  get pageLengthOptions(): number[] { return this._pageLengthOptions; }
+  _pageLengthOptions: number[] = [];
+
+  /** Event emitted when the paginator changes the page length or current page index. */
+  @Output() pageChange = new EventEmitter<PageChangeEvent>();
+
+  /** Displayed set of page length options. Will be sorted and include current page length. */
+  _displayedPageLengthOptions: number[];
+
+  constructor(public _intl: MdPaginatorIntl) { }
+
+  ngOnInit() {
+    this._initialized = true;
+
+    // If no page length or options have been provided, use the list length as the only option.
+    // Otherwise if there is no page length but there are options, then use the first option.
+    if (this.pageLength == undefined) {
+      this.pageLength = this.pageLengthOptions.length != 0 ?
+          this.pageLengthOptions[0] :
+          this.listLength;
+    }
+
+    this._updateDisplayedPageLengthOptions();
+  }
+
+  /** Increments the current page index to the next page index if a next page exists. */
+  navigateToNextPage() {
+    if (!this.canNavigateToNextPage()) { return; }
+    this.currentPageIndex++;
```

Comment:
Right now we only do fire on interaction (user selects a different page size or navigated between pages). We could consider defaulting pageLength to 0 and removing this check to default to a certain page size. I believe that would make this more stateless.